File: Linear_Control_Course/Chapter27/Lesson1/Chapter27_Lesson1.py
import numpy as np
import matplotlib.pyplot as plt
import control as ctl  # python-control library

# Plant and controller parameters
tau = 0.5
G = ctl.tf([1.0], [tau, 1.0])   # G(s) = 1 / (tau s + 1)
Kp = 4.0
Ki = 5.0
C = ctl.tf([Kp, Ki], [1.0, 0.0])  # Kp + Ki / s

L = C * G
S = 1 / (1 + L)
T = L / (1 + L)

# 1) Tracking: step in reference, no disturbance
t = np.linspace(0, 5, 500)
t1, y_track = ctl.step_response(T, T=t)  # y(t) for r(t) = 1

# 2) Regulation: step disturbance at plant output, r(t) = 0
# For output disturbance d(t), transfer from d to y is S(s).
t2, y_reg = ctl.step_response(S, T=t)

plt.figure()
plt.plot(t1, y_track, label="Tracking: y(t) for step r(t)")
plt.plot(t2, y_reg, label="Regulation: y(t) for step d(t)")
plt.axhline(1.0, linestyle="--", label="Reference level r(t) = 1")
plt.axhline(0.0, linestyle=":", label="Regulation target y(t) = 0")
plt.xlabel("Time [s]")
plt.ylabel("Output y(t)")
plt.legend()
plt.grid(True)
plt.title("Tracking vs Regulation on First-Order Plant")
plt.show()

File: Tutorials/Codes/Linear_Control_Course/Chapter27/Lesson1/Chapter27_Lesson1.java
public class TrackingVsRegulationDemo {

    public static void main(String[] args) {
        double tau = 0.5;
        double Kp  = 4.0;
        double Ki  = 5.0;

        double h   = 0.001;
        double T   = 5.0;
        int N      = (int) (T / h);

        double[] t      = new double[N];
        double[] yTrack = new double[N];
        double[] yReg   = new double[N];

        // Tracking scenario
        double x = 0.0;
        double u = 0.0;
        double ePrev = 0.0;

        for (int k = 0; k < N; ++k) {
            double tk = k * h;
            t[k] = tk;

            double r = 1.0;
            double d = 0.0;

            double y = x + d;
            double e = r - y;

            double du = Kp * (e - ePrev) + Ki * h * e;
            u += du;
            ePrev = e;

            double dx = (-1.0 / tau) * x + (1.0 / tau) * u;
            x += h * dx;

            yTrack[k] = y;
        }

        // Regulation scenario
        x = 0.0;
        u = 0.0;
        ePrev = 0.0;

        for (int k = 0; k < N; ++k) {
            double tk = k * h;

            double r = 0.0;
            double d = (tk >= 1.0) ? 0.5 : 0.0;

            double y = x + d;
            double e = r - y;

            double du = Kp * (e - ePrev) + Ki * h * e;
            u += du;
            ePrev = e;

            double dx = (-1.0 / tau) * x + (1.0 / tau) * u;
            x += h * dx;

            yReg[k] = y;
        }

        // Print a few samples
        for (int k = 0; k < N; k += N / 10) {
            System.out.println(String.format(
                "t = %.3f, yTrack = %.3f, yReg = %.3f",
                t[k], yTrack[k], yReg[k]
            ));
        }
    }
}
